src/storage/redis.rs

```diff
@@ -0,0 +1,400 @@
+                    err.to_string(),
+                ))),
+            }
+        } else {
+            Err(RedisError::from((
+                ErrorKind::TypeError,
+                "expected a data type from redis",
+            )))
+        }
+    }
+}
+
+#[async_trait]
+impl Storage for RedisStorage {
+    async fn get(&self, id: &str) -> Result<Option<Message>> {
+        let mut conn = self.get_connection().await?;
+        let key = self.prefixed(Queue::Backlog(id));
+        let ret: Option<Vec<u8>> = conn.get(key).await?;
+
+        match ret {
+            Some(val) => {
+                let msg = Message::from_json(&val)?;
+                Ok(Some(msg))
+            }
+            None => Ok(None),
+        }
+    }
+
+    async fn run(&self, msg: &Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let queue = self.prefixed(Queue::Run(&msg.command));
+
+        conn.rpush(&queue, msg).await?;
+
+        // only keep `max_commands` in this queue
+        let len: isize = conn.llen(&queue).await?;
+        if len > self.max_commands {
+            conn.ltrim(&queue, self.max_commands, -1).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn forward(&self, msg: &Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+
+        // add to backlog
+        let key = self.prefixed(Queue::Backlog(&msg.id));
+        conn.set_ex(&key, msg, self.ttl as usize).await?;
+
+        // push to forward for every destination
+        let queue = self.prefixed(Queue::Forward);
+        for destination in &msg.destination {
+            let forwarded = ForwardedMessage {
+                id: msg.id.to_owned(),
+                destination: *destination,
+            };
+            conn.rpush(&queue, &forwarded).await?
+        }
+
+        Ok(())
+    }
+
+    async fn reply(&self, msg: &Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+
+        conn.rpush(&msg.reply, msg).await?;
+        Ok(())
+    }
+
+    async fn local(&self) -> Result<Message> {
+        let mut conn = self.get_connection().await?;
+        let queue = self.prefixed(Queue::Local);
+        let ret: (Vec<u8>, Vec<u8>) = conn.blpop(&queue, 0).await?;
+
+        let (_, value) = ret;
+        let msg = Message::from_json(&value)?;
+        Ok(msg)
+    }
+
+    async fn queued(&self) -> Result<QueuedMessage> {
+        let mut conn = self.get_connection().await?;
+        let forward_queue = self.prefixed(Queue::Forward);
+        let reply_queue = self.prefixed(Queue::Reply);
+        let queues = (forward_queue, reply_queue);
+
+        let result;
+
+        loop {
+            let ret: (String, Vec<u8>) = conn.blpop(&queues, 0).await?;
+            let (queue, value) = ret;
+
+            match queue {
+                forward_queue => {
+                    let forwarded = ForwardedMessage::from_bytes_pair(&value)?;
```

Comment:
i really think this section can be improved code wise. 
for example, use the from_redis_value directly since it's implemenbted on the ForwardMessage type.

```suggestion
                    let forwarded = ForwardedMessage::from_redis_value(&value);
```

Reply:
You also shouldn't do `?` here because what if someone polluted the queue with wrong unparasable values ? instead of the value is not parsable, u log the error and try again.